bpf: fix __snat_clamp_port_range not using highest port

lmb opened the pull request:
bpf: fix nodeport asserts

    The nodeport code has a couple of configurable port ranges. In the code they
    are used as closed ranges [min, max] but the asserts do not allow min ==
    max, aka the interval containing just min.

    Fixes: d3b6298de7 ("bpf: add build assertions for nodeport assumptions") 
    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

bpf: fix __snat_clamp_port_range not using highest port

    __snat_clamp_port_range has an output in the range [start, end), but the
    call sites expect it to produce a value in [start, end].

    Operate on the full range and fix tests which relied on the
    [start, end) interpretation.

    Fixes: 92cd032bf7 ("bpf, snat: select lru map if available otherwise fall
    back to htab") Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

bpf: avoid modulo in __snat_clamp_port

    Modulo is an expensive operation. Replace it with an integer multiplication
    adapted from bounded random number generation.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

lmb committed:
bpf: fix nodeport asserts

The nodeport code has a couple of configurable port ranges. In the
code they are used as closed ranges [min, max] but the asserts
do not allow min == max, aka the interval containing just min.

Fixes: d3b6298de7 ("bpf: add build assertions for nodeport assumptions")
Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

gentoo-root commented on bpf/lib/nat.h:
```diff
@@ -40,12 +40,26 @@ struct nat_entry {
+ *
+ * Introduces a slight bias.
+ *
+ * Adapted from "Integer Multiplication (Biased)" in https://www.pcg-random.org/posts/bounded-rands.html
```
Interesting overview of bounding methods!

gentoo-root commented on bpf/lib/nat.h:
```diff
@@ -40,12 +40,26 @@ struct nat_entry {
 {
-	return (val % (__u16)(end - start)) + start;
+	/* Account for closed interval. */
+	__u32 n = (__u32)(end - start) + 1;
```
The `__u32` cast should not be necessary due to integer promotion (__u16 will be converted to int before doing the subtraction), but the code looks more straightforward with it, let's keep it.

gentoo-root commented on bpf/lib/nat.h:
```diff
@@ -40,12 +40,26 @@ struct nat_entry {
 
+/* Clamp a port to the range [start, end].
+ *
+ * Introduces a slight bias.
```
It will be a slight bias if `val` is evenly distributed over the whole u16 range. But as far as I understand, it's not, so the bias will be higher — or am I wrong?

I'd prefer if you split the off-by-one fix and the switch to the new method into two separate commits.

lmb replied:
> But as far as I understand, it's not, so the bias will be higher — or am I wrong?

You mean for the case where we try to keep the original port? I don't follow, sorry.

> I'd prefer if you split the off-by-one fix and the switch to the new method into two separate commits.

I'll give it a try!

gentoo-root replied:
> You mean for the case where we try to keep the original port? I don't follow, sorry.

Sorry that I wasn't clear from the start, let me bring in some more details:

__snat_try_keep_port calls __snat_clamp_port_range only if val is not in [start, end] (yes, basically when we _can't_ keep the original port). It limits the effective range of val in __snat_try_keep_port to [1, start) \cup (end, 65535]. The algorithm you used produces numbers with a slight bias only if val is uniform over the whole range [0, 65535]. It's not our case, so the bias will be huge, basically, certain outputs will never occur: from approx. `start*n / 2^16` to approx. `end*n / 2^16`.

To alleviate it, we could first check whether val \in [1, start) or val \in (end, 65535], then apply the algorithm to one of those subranges, but the problem now is that the interval length is not a power of two (unless start and end are chosen "good"), so effective division is not possible.

lmb replied:
But in __snat_clamp_port_range val is always a random number, not the original port (except in the weird retry cases, where we do port+1)?

msune replied:
@lmb @gentoo-root any progress on this?

gentoo-root replied:
Sorry, I somehow missed Lorenz's reply =/

> in __snat_clamp_port_range val is always a random number

Yes, you are right, I thought we used the same val. Thanks for the clarification, works for me!

lmb committed:
bpf: avoid modulo in __snat_clamp_port

Modulo is an expensive operation. Replace it with an integer
multiplication adapted from bounded random number generation.

Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

lmb committed:
bpf: fix __snat_clamp_port_range not using highest port

__snat_clamp_port_range has an output in the range [start, end),
but the call sites expect it to produce a value in [start, end].

Operate on the full range and fix tests which relied on the
[start, end) interpretation.

Fixes: 92cd032bf7 ("bpf, snat: select lru map if available otherwise fall back to htab")
Signed-off-by: Lorenz Bauer <lmb@isovalent.com>
